Add Cavity Flow Navier-Stokes notebook to cfd examples

dabiged opened the pull request:
This commit adds a functioning implementation of the Navier-Stokes equations in the cavity flow regime, using both the numpy version from "12 steps to Navier-Stokes" (note: Coordinates have been changed to x,y from y,x) and a working devito version.

mloubout commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,566 @@
+{
```
<ul><li><code>first_derivative(pn, dim=y)</code>  are these debug leftovers?</li><li><code>u.data[:] = 0.</code>  and such are not necessary, initialized to zero by default</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

dabiged replied:
>     * `first_derivative(pn, dim=y)`  are these debug leftovers?
Yes. I didn't know you could do p.dx or p.dxc when I started. I will remove these.
>     * `u.data[:] = 0.`  and such are not necessary, initialized to zero by default
I didn't know this happened

mloubout commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,566 @@
+{
```
<p>Can fix the error with an epsilon:</p><pre class="ql-syntax" spellcheck="false">l1norm = (np.sum(np.abs(_p.data[:]) - np.abs(_pn.data[:])) /
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; (np.sum(np.abs(_p.data[:])) + 1e-6))

</pre><p>You also can use l1-norm in numpy via <code>np.norm(_p.data[:], 1)</code></p><p> </p><p><code>oppres(p=_p, pn=_pn,time=3)</code>  any reason for <code>time_3</code>  or just empirical?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

dabiged replied:
I subsequently had 
` if step not 0: `

I might switch to the np.norm.

time_3 is a debug leftover. The solutions wouldn't converge and I thought taking more smaller steps would help. I didn't.

mloubout commented:
Super Cool!!!! Left some minor comments but looks really nice

dabiged commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,566 @@
+    "u_dx = first_derivative(u, dim=x)\n",
+    "u_dy = first_derivative(u, dim=y)\n",
+    "\n",
+    "v = TimeFunction(name='v', grid=grid, space_order=2)\n",
+    "v_dx = first_derivative(v, dim=x)\n",
+    "v_dy = first_derivative(v, dim=y)\n",
```
Should I have a comment here highlighting this? This was explicitly requested the other day in slack.

mloubout replied:
Yes please add a comment, I personally think this should be removed since the shortcuts (`.dx, .dy, ...`) are used for the equations.

georgebisbas commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,566 @@
+{
```
<pre class="ql-syntax" spellcheck="false">&lt;ipython-input-12-e539ef5b065a&gt;:34: RuntimeWarning: invalid value encountered in true_divide
  l1norm = (np.sum(np.abs(_p.data[:]) - np.abs(_pn.data[:])) /
</pre><p>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

dabiged replied:
The initial solution to the pressures has everything as zero. This means you divide by zero on the first time step. 
This can be avoided either by skipping the pressure solve on the first step, or adding epsilon to the denominator.

georgebisbas commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,551 @@
+{
```
<p>Should we execute for this error? Maybe avoid it?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

georgebisbas commented on examples/cfd/08_channel_flow.ipynb:
```diff
@@ -0,0 +1,549 @@
+{
```
<p>#Silence non-essential outputs from the solver.</p><p> </p><p>We need to add space after the comment.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

dabiged replied:
Fixed.

georgebisbas commented on examples/cfd/08_channel_flow.ipynb:
```diff
@@ -0,0 +1,549 @@
+{
```
<p>Capital in comments please.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

dabiged commented:
Hi All, I just pushed a new commit of this. It include changes with respect to all the comments in here. I have also fixed the coordinate system in the numpy section to be the same as the other cfd notebooks.

mloubout commented on examples/cfd/08_channel_flow.ipynb:
```diff
@@ -0,0 +1,541 @@
+    "\n",
+    "v = TimeFunction(name='v', grid=grid, space_order=2)\n",
+    "v_dx = first_derivative(v, dim=x)\n",
+    "v_dy = first_derivative(v, dim=y)\n",
```
would remove here to.

mloubout commented on examples/cfd/08_channel_flow.ipynb:
```diff
@@ -0,0 +1,541 @@
+   "source": [
+    "#NBVAL_IGNORE_OUTPUT\n",
+    "fig = pyplot.figure(figsize = (11,7), dpi=100)\n",
+    "pyplot.quiver(X, Y, u, v);"
```
Can we have the contour of the pressure as well here for comparison with the devito result.

dabiged commented:
Removed lines using explicit `derivative` calls.
Replaced calls to these objects with `function.dxc` or `function.dxy`
Added pressure contour plot to channel flow notebook part a. I am unable to add contour lines as the pressure is constant and attempting to derive countour lines results in a UserError.

mloubout commented:
>. I am unable to add contour lines as the pressure is constant and attempting to derive countour lines results in a UserError.
Wait, so the result is different with devito?

rhodrin commented:
@dabiged Would you mind also adding some comparison of the results with lab data? (Example below). Thanks.

```
# Check results
Marchi_Re10_u = np.array([[0.0625, -3.85425800e-2],
                          [0.125,  -6.96238561e-2],
                          [0.1875, -9.6983962e-2],
                          [0.25,  -1.22721979e-1],
                          [0.3125, -1.47636199e-1],
                          [0.375,  -1.71260757e-1],
                          [0.4375, -1.91677043e-1],
                          [0.5,    -2.05164738e-1],
                          [0.5625, -2.05770198e-1],
                          [0.625,  -1.84928116e-1],
                          [0.6875, -1.313892353e-1],
                          [0.75,   -3.1879308e-2],
                          [0.8125,  1.26912095e-1],
                          [0.875,   3.54430364e-1],
                          [0.9375,  6.50529292e-1]])

Marchi_Re10_v = np.array([[0.0625, 9.2970121e-2],
                          [0.125, 1.52547843e-1],
                          [0.1875, 1.78781456e-1],
                          [0.25, 1.76415100e-1],
                          [0.3125, 1.52055820e-1],
                          [0.375, 1.121477612e-1],
                          [0.4375, 6.21048147e-2],
                          [0.5, 6.3603620e-3],
                          [0.5625, -5.10417285e-2],
                          [0.625, -1.056157259e-1],
                          [0.6875, -1.51622101e-1],
                          [0.75, -1.81633561e-1],
                          [0.8125, -1.87021651e-1],
                          [0.875, -1.59898186e-1],
                          [0.9375, -9.6409942e-2]])

x_coord = np.linspace(0, Lx, grid.shape[0])
y_coord = np.linspace(0, Ly, grid.shape[1])

fig = plt.figure(figsize=(12, 6))
ax1 = fig.add_subplot(121)
ax1.plot(u.data[-1,int(grid.shape[0]/2),:],y_coord[:])
ax1.plot(Marchi_Re10_u[:,1],Marchi_Re10_u[:,0],'ro')
ax1.set_xlabel('$u$')
ax1.set_ylabel('$y$')
ax1 = fig.add_subplot(122)
ax1.plot(x_coord[:],v.data[-1,:,int(grid.shape[0]/2)])
ax1.plot(Marchi_Re10_v[:,0],Marchi_Re10_v[:,1],'ro')
ax1.set_xlabel('$x$')
ax1.set_ylabel('$v$')

plt.show()
```

rhodrin commented:
Also, for the second notebook do you have any validity check? Would you mind also adding an assert ensuring that `norm(pure_python-devito)<some_suitable_tolerance`?

rhodrin commented:
@dabiged Thanks for adding the plots. It would seem the flow has not yet reached steady state though - would you be able to increase the simulation time so that it runs until it has reached steady state?

After that, and the addition of a couple of asserts (ensuring pure python vs devito results are the same to a reasonable tolerance) it should be GTG.

Thanks again for your work on this.

georgebisbas commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,642 @@
+{
```
<p>What about </p><p><code>import numpy as np</code></p><p> </p><p>and then tweak accordingly ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

georgebisbas commented on examples/cfd/08_channel_flow.ipynb:
```diff
@@ -0,0 +1,543 @@
+{
```
<p>##variable declarations</p><p>==&gt; ## Variable declarations</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

georgebisbas replied:
<p>import numpy is redundant here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

dabiged commented:
Latest commit fixes NB 07 which I think is now ready to merge. I have incorporated @rhodrin 's suggestion about the Marchi et al paper more formally and fixed all the issues raised by @georgebisbas (numpy imports and comments not conforming to standards)

NB 08 is another matter entirely. The periodic boundary condition is still not working correctly in the devito implementation. I have added another computation below the current one to a later point in time to show that this computation is not stable (This breaks the notebook badly and I am not sure if this is right etiquette for Pull requests). I do not know what I am doing wrong here and I am open to suggestions. 
@mloubout suggested altering the op.apply(x_m=1, x_M=nx-2) or similar. I am open to trying something like this but I don't know how best to proceed with this PR. Should I try and implement this solution and push a new commit with it?

rhodrin commented:
@dabiged Thanks for the update.

Sorry to be a pain, but a couple of comments/questions: 
- I only just noticed that your domain is 2x2 sorry, for the BC's you're using the Machi results are more suited to a 1x1 domain - I'm not sure what the Re is in your case? (For a 1x1 domain you should see the results match up more closely). Any objections to using a 1x1 domain or adjusting the bc appropriately or comparing with results at a different Re? Sorry again about that, I should have noticed this difference before. (Also the Machi plot axes are currently for a 1x1 domain).
- The Python vs Devito results seem quite a bit different? Any idea why this?

Re. notebook 8, how would you like to proceed with the PRs? If you like you could remove nb 8 from this PR and just merge 7 for the time being, and submit 8 at a later date in a separate PR? Or hang on until notebook 8 is also complete? And regarding that, we are organizing a sprint for writing the [Devito Book](https://github.com/devitocodes/devito_book) which will take place on the 30th and 31st of this month. If you happen to be interested in joining this event (more info will be appearing on our slack channels soon), and since we will require periodic boundary condition implementations for various problems in the book, I'd be happy to look closely at the Devito implementation of the problem in notebook 8 with you if you'd like.

georgebisbas commented on examples/cfd/07_cavity_flow.ipynb:
```diff
@@ -0,0 +1,692 @@
+{
```
<p>LInk for that table? Ah ok, never mind.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1368/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

rhodrin commented:
Superseded by #1502.
